File: ruby/lib/idkollen/age_verification.rb
# frozen_string_literal: true

module Idkollen
  module AgeVerification
    class Request
      attr_reader :min_age, :max_age, :ref_id, :callback_url, :redirect_url

      def initialize(min_age: nil, max_age: nil, ref_id: nil, callback_url: nil, redirect_url: nil)
        @min_age = min_age
        @max_age = max_age
        @ref_id = ref_id
        @callback_url = callback_url
        @redirect_url = redirect_url
      end

      def to_h
        {
          minAge: @min_age,
          maxAge: @max_age,
          refId: @ref_id,
          callbackUrl: @callback_url,
          redirectUrl: @redirect_url,
        }.compact
      end
    end

    class Pending
      attr_reader :id, :url, :min_age, :max_age

      def initialize(id:, url: nil, min_age: nil, max_age: nil)
        @id = id; @url = url; @min_age = min_age; @max_age = max_age
      end

      def deconstruct_keys(_) = { id: @id, url: @url, min_age: @min_age, max_age: @max_age }

      def self.from_json(d)
        new(id: d['id'], url: d['url'], min_age: d['minAge'], max_age: d['maxAge'])
      end
    end

    class Completed
      attr_reader :id, :age_verified

      def initialize(id:, age_verified:)
        @id = id; @age_verified = age_verified
      end

      def deconstruct_keys(_) = { id: @id, age_verified: @age_verified }

      def self.from_json(d) = new(id: d['id'], age_verified: d['ageVerified'])
    end

    class Failed
      attr_reader :id, :error

      def initialize(id:, error:)
        @id = id; @error = error
      end

      def deconstruct_keys(_) = { id: @id, error: @error }

      def self.from_json(d) = new(id: d['id'], error: d['error'])
    end

    module Status
      def self.from_json(d)
        case d['status']
        when 'PENDING'   then Pending.from_json(d)
        when 'COMPLETED' then Completed.from_json(d)
        when 'FAILED'    then Failed.from_json(d)
        else raise ArgumentError, "Unknown age verification status: #{d['status']}"
        end
      end
    end
  end
end

File: swift/Sources/IdkollenClient/AgeVerification.swift
import Foundation

public struct AgeVerificationRequest: Encodable {
    public let minAge: Int?
    public let maxAge: Int?
    public let refId: String?
    public let callbackUrl: String?
    public let redirectUrl: String?

    public init(minAge: Int? = nil, maxAge: Int? = nil, refId: String? = nil, callbackUrl: String? = nil, redirectUrl: String? = nil) {
        self.minAge = minAge
        self.maxAge = maxAge
        self.refId = refId
        self.callbackUrl = callbackUrl
        self.redirectUrl = redirectUrl
    }
}

public enum AgeVerificationStatus: Decodable {
    case pending(AgeVerificationPending)
    case completed(AgeVerificationCompleted)
    case failed(AgeVerificationFailed)

    private enum DiscriminatorKey: String, CodingKey { case status }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: DiscriminatorKey.self)
        let status = try container.decode(String.self, forKey: .status)
        switch status {
        case "PENDING":   self = .pending(try AgeVerificationPending(from: decoder))
        case "COMPLETED": self = .completed(try AgeVerificationCompleted(from: decoder))
        case "FAILED":    self = .failed(try AgeVerificationFailed(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: .status, in: container, debugDescription: "Unknown age verification status: \(status)")
        }
    }
}

public struct AgeVerificationPending: Decodable {
    public let id: String
    public let url: String?
    public let minAge: Int?
    public let maxAge: Int?
}

public struct AgeVerificationCompleted: Decodable {
    public let id: String
    public let ageVerified: Bool
}

public struct AgeVerificationFailed: Decodable {
    public let id: String
    public let error: String
}
